### PR TITLE
Update LLMs fine-tuning with Ray example for Llama 3.1 models

### PR DESCRIPTION
**Pull request by astefanutti**
## Description

This PR updates the Ray fine-tuning example so it works for Llama 3.1 model series.

## How Has This Been Tested?

The changes in this PR have been successfully tested with the following models:

* meta-llama/Meta-Llama-3.1-8B (red)
* meta-llama/Meta-Llama-3.1-8B-Instruct (light blue)
* meta-llama/Llama-2-7b-chat-hf (orange and dark blue)

image

## Merge criteria:
- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: examples/ray-finetune-llm-deepspeed/README.md [hunk 1]
```diff
@@ -78,6 +76,37 @@ By default, at the end of each epoch, the checkpoints are stored in the configur
...
+### Llama 3 8B - GSM8k - LoRA
+
+* OpenShift cluster:
+  * ROSA-hosted 4.14.20
+  * 5 `g5.8xlarge` (A10 Nvidia GPU) worker nodes
+* Ray cluster:
+    ```python
+    ClusterConfiguration(
+        num_workers=4,
+        min_cpus=8,
```
Comment (Bobbins228) [inline examples/ray-finetune-llm-deepspeed/README.md:88]:
```suggestion
        worker_cpu_requests=8,
```

File: examples/ray-finetune-llm-deepspeed/README.md [hunk 2]
```diff
@@ -78,6 +76,37 @@ By default, at the end of each epoch, the checkpoints are stored in the configur
...
+        max_cpus=16,
```
Comment (Bobbins228) [inline examples/ray-finetune-llm-deepspeed/README.md:89]:
```suggestion
        worker_cpu_limits=16,
```

File: examples/ray-finetune-llm-deepspeed/README.md [hunk 3]
```diff
@@ -78,6 +76,37 @@ By default, at the end of each epoch, the checkpoints are stored in the configur
...
+        head_cpus=8,
+        min_memory=32,
```
Comment (Bobbins228) [inline examples/ray-finetune-llm-deepspeed/README.md:91]:
```suggestion
        worker_memory_requests=32,
```

File: examples/ray-finetune-llm-deepspeed/README.md [hunk 4]
```diff
@@ -78,6 +76,37 @@ By default, at the end of each epoch, the checkpoints are stored in the configur
...
+        max_memory=64,
```
Comment (Bobbins228) [inline examples/ray-finetune-llm-deepspeed/README.md:92]:
```suggestion
        worker_memory_limits=64,
```

File: examples/ray-finetune-llm-deepspeed/README.md [hunk 5]
```diff
@@ -78,6 +76,37 @@ By default, at the end of each epoch, the checkpoints are stored in the configur
...
+        head_memory=64,
+        head_gpus=1,
```
Comment (Bobbins228) [inline examples/ray-finetune-llm-deepspeed/README.md:94]:
```suggestion
        head_extended_resource_requests={"nvidia.com/gpu": 1},
```

File: examples/ray-finetune-llm-deepspeed/README.md [hunk 6]
```diff
@@ -78,6 +76,37 @@ By default, at the end of each epoch, the checkpoints are stored in the configur
...
+        num_gpus=1,
```
Comment (Bobbins228) [inline examples/ray-finetune-llm-deepspeed/README.md:95]:
```suggestion
        worker_extended_resource_requests={"nvidia.com/gpu": 1},
```

**Review by Bobbins228 — CHANGES_REQUESTED** [review]
Just some nits for the Cluster Configuration parameters that are being depreciated. - The notebook should be updated with these params as well :) 

I am going to try and test this today

**Comment by astefanutti** [comment]
> Just some nits for the Cluster Configuration parameters that are being depreciated. - The notebook should be updated with these params as well :)
> 
> I am going to try and test this today

Would you know the timeline when these new fields got introduced? If they'll all be available in the upcoming RHOAI 2.12 release, I guess we can make upgrading to that version a prerequisite.

**Comment by Bobbins228** [comment]
@astefanutti These fields are introduced as of SDK v0.18.0.

File: examples/ray-finetune-llm-deepspeed/README.md [hunk 7]
```diff
@@ -78,6 +76,37 @@ By default, at the end of each epoch, the checkpoints are stored in the configur
...
+* Ray job:
+    ```python
+    ray_finetune_llm_deepspeed.py "
+        "--model-name=meta-llama/Meta-Llama-3.1-8B "
+        "--ds-config=./deepspeed_configs/zero_3_offload_optim+param.json "
+        f"--storage-path=s3://{s3_bucket}/ray_finetune_llm_deepspeed/ "
+        "--lora "
+        "--num-devices=5 "
+        "--batch-size-per-device=6 "
+        "--eval-batch-size-per-device=8 "
```
Comment (ChristianZaccaria) [inline examples/ray-finetune-llm-deepspeed/README.md:107]: For my own knowledge: Is there a 'rule-of-thumb' to follow for these values in relation to the number of devices that are available? Or is it a case of trial and error based on how it is performing?

**Comment by ChristianZaccaria** [comment]
> @astefanutti These fields are introduced as of SDK v0.18.0.

Correct ^, that was 3 weeks ago. However, I discovered just now that the `odh-notebooks-sync` workflow hasn't been working for around 2 months, the latest SDK available in RHOAI notebook images is v0.16.0, hence the new parameters will probably not be included for 2.12.
https://github.com/project-codeflare/codeflare-sdk/actions/workflows/odh-notebooks-sync.yml

File: examples/ray-finetune-llm-deepspeed/README.md (hunk 7 above)
Comment (astefanutti) [reply examples/ray-finetune-llm-deepspeed/README.md:107]: Deepspeed has many configuration parameters that affects how much memory is required w.r.t. the number of devices. It provides some tools to estimate the memory requirements depending on the configuration: https://deepspeed.readthedocs.io/en/latest/memory.html. Beyond, those estimates, that remains mostly empirical.

Comment (ChristianZaccaria) [reply examples/ray-finetune-llm-deepspeed/README.md:107]: Thanks!! - That's super helpful

**Comment by ChristianZaccaria** [comment]
I'm checking with Harshad for which rhoai release will the new SDK (0.19.1) be available. In any case, you're right, we are still using i.e., `num_gpus` in our notebooks. I think we can keep it as such until we update everything.

**Comment by astefanutti** [comment]
@Bobbins228 I've updated it to use the latest SDK version (and rebased on top of what @Ygnas had already updated). Could you please take a look?

**Comment by astefanutti** [comment]
@sutaakar @abhijeet-dhumal is it OK to merge this or there is anything that requires to be adjusted from the testing side?

**Comment by sutaakar** [comment]
It looks fine to me, though I didn't run it by myself.
Can be verified as part of test coverage done by Abhijeet.

**Comment by Bobbins228** [comment]
@astefanutti /lgtm 
I can verify it myself on top of Abhijeets test coverage

**Comment by Bobbins228** [comment]
I was unsuccessful in gathering GPU instances. Spoke with Abhijeet and he is currently testing the new example, he will drop an update here when finished.

**Comment by abhijeet-dhumal** [comment]
@astefanutti @Bobbins228 
I ran test by rebasing [my test-PR](https://github.com/opendatahub-io/distributed-workloads/pull/197/commits/304c810496466fdef417310addc84ede53fad52d) with this PR using `zero_3_offload_optim+param.json` and did some changes in [input params](https://github.com/opendatahub-io/distributed-workloads/pull/197/commits/304c810496466fdef417310addc84ede53fad52d#diff-518bd56270f4732788268a6aeb8f934c77f8d4436ae0aa2961e434bc54be1cd0R60) but it raises below error each time after 7 mins during  rayJobExecution : 
```
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 7.81 GiB. GPU
```
[ray-job-log-raysubmit_r7HQfNmc1K1RTw4E.log](https://github.com/user-attachments/files/16750933/ray-job-log-raysubmit_r7HQfNmc1K1RTw4E.log)

but when I ran same test using `zero_3_llama_2_7b.json`, it works seamlessly .. 
It seems there are no changes required in this PR from testing perspective ! :rocket:

**Comment by astefanutti** [comment]
@abhijeet-dhumal Llama 3.1 8B has more parameters than Llama 2 7B so it requires more memory. You can try lowering the batch size using the `batch-size-per-device` option, otherwise you can explicitly set the pre-trained model to Llama 2 7B which should restore your previous configuration.

**Comment by abhijeet-dhumal** [comment]
Added test for Llama 3.1 8B :  https://github.com/opendatahub-io/distributed-workloads/pull/197/files#diff-518bd56270f4732788268a6aeb8f934c77f8d4436ae0aa2961e434bc54be1cd0R35

Test-Configuration : 
1. 1x g5.8xlarge gpu
2. ModelConfig : zero_3_llama_2_7b.json / zero_3_offload_optim_param.json
3. head_cpus=2
4. num_workers=1
5. head_memory=48
6. batch-size-per-device = 6
7. worker_memory_requests=64
8. worker_memory_limits=128

image (124)

**Comment by astefanutti** [comment]
@abhijeet-dhumal I've renamed the file so it doesn't have the problematic character. PTAL.

**Review by sutaakar — APPROVED** [review]

**Pull request merged**